jobspec2/hcl_conversions.go

```diff
@@ -291,20 +302,102 @@ func decodeTaskGroup(body hcl.Body, ctx *hcl.EvalContext, val interface{}) hcl.D
 func decodeTask(body hcl.Body, ctx *hcl.EvalContext, val interface{}) hcl.Diagnostics {
 	// special case scaling policy
 	t := val.(*api.Task)
-	b, remain, diags := body.PartialContent(&hcl.BodySchema{
+
+	var diags hcl.Diagnostics
+
+	// special case env and meta
+	envAttr, body, moreDiags := decodeAsAttribute(body, ctx, "env")
+	diags = append(diags, moreDiags...)
+	metaAttr, body, moreDiags := decodeAsAttribute(body, ctx, "meta")
+	diags = append(diags, moreDiags...)
+
+	b, remain, moreDiags := body.PartialContent(&hcl.BodySchema{
 		Blocks: []hcl.BlockHeaderSchema{
 			{Type: "scaling", LabelNames: []string{"name"}},
 		},
 	})
 
+	diags = append(diags, moreDiags...)
 	diags = append(diags, decodeTaskScalingPolicies(b.Blocks, ctx, t)...)
 
 	decoder := newHCLDecoder()
 	diags = append(diags, decoder.DecodeBody(remain, ctx, val)...)
 
+	if envAttr != nil {
+		t.Env = envAttr
+	}
+	if metaAttr != nil {
+		t.Meta = metaAttr
+	}
+
 	return diags
 }
 
+// decodeAsAttribute decodes the named field as an attribute assignment if found.
+//
+// Nomad jobs contain attributes (e.g. `env`, `meta`) that are meant to contain arbitrary
+// keys. HCLv1 allowed both block syntax (the preferred and documented one) as well as attribute
+// assignment syntax:
+//
+// ```hcl
+// # block assignment
+// env {
+//   ENV = "production"
+// }
+//
+// # as attribute
+// env = { ENV: "production" }
+// ```
+//
+// HCLv2 block syntax, though, restricts valid input and doesn't allow dots or invalid identifiers
+// as block attribute keys.
+// Thus, we support both syntax to unrestrict users.
+//
+// This function attempts to read the named field, as an attribute, and returns
+// found map, the remaining body and diagnostics. If the named field is found
+// with block syntax, it returns a nil map, and caller falls back to reading
+// with block syntax.
+//
+func decodeAsAttribute(body hcl.Body, ctx *hcl.EvalContext, name string) (map[string]string, hcl.Body, hcl.Diagnostics) {
+	b, remain, diags := body.PartialContent(&hcl.BodySchema{
+		Attributes: []hcl.AttributeSchema{
+			{Name: name, Required: false},
+		},
+	})
+
+	if diags.HasErrors() || b.Attributes[name] == nil {
+		// ignoring errors, to avoid duplicate errors. True errors will
+		// reported in the fallback path
+		return nil, body, nil
```

Comment:
I don't think ignoring the errors is sound always, but don't know of a better heaurstic.

If we get an error when parsing as an attribute but no error when parsing as block, nomad shouldn't report any error.

However, if the job fails to parse both ways, one error might be more insightful than the other.  Here, we use the parse-as-block error, which may be less ideal; but it's not clear how we can decide which error is better.

Reply:
agreed, i don't think we can decide which is better without knowing which they were attempting. in that case, it might be better to print both errors, which a sort of "neither of these parsing approaches worked, we aren't sure which you were attempting, here are the errors from each"

Reply:
I'm not totally sure I know which errors can bubble up from this, but supposing we have an attribute like this:

```hcl
env = {
  "my.env" = local.non_existing_local // this local doesn't exist so throws an error
}
```

Won't this produce a misleading error about the `.` in the key when we try to parse it as a block? Maybe a test example of what to expect here would be helpful.

Reply:
I'll add a test.  @tgross In this case, the error for undefined local will be returned in `DecodeExpression` call.  The diags in this line should only be about the partial schema conformance for `env` attribute.

Reply:
Updated with invalid cases tests in 2ae47613566f6aa530b33e1c6fdd279a5a96e3c6 .